src/coreclr/src/gc/gc.cpp

```diff
@@ -28045,182 +27942,173 @@ void gc_heap::revisit_written_pages (BOOL concurrent_p, BOOL reset_only_p)
+                                    revisit_written_page (page, high_address, concurrent_p,
+                                                        seg, last_page, last_object,
+                                                        !small_object_segments,
+                                                        total_marked_objects);
+                                }
+                                else
+                                {
+                                    dprintf (3, ("page %d at %Ix is >= %Ix!", i, (size_t)page, (size_t)high_address));
+                                    assert (!"page shouldn't have exceeded limit");
+                                }
                             }
                         }
-                    }
 
-                    if (bcount >= array_size){
-                        base_address = background_written_addresses [array_size-1] + WRITE_WATCH_UNIT_SIZE;
-                        bcount = array_size;
+                        if (bcount >= array_size){
+                            base_address = background_written_addresses [array_size-1] + WRITE_WATCH_UNIT_SIZE;
+                            bcount = array_size;
+                        }
+                    }
+                    else
+                    {
+                        break;
                     }
-                }
-                else
-                {
-                    break;
                 }
             }
+
+            seg = heap_segment_next_rw (seg);
         }
 
-        seg = heap_segment_next_rw (seg);
+        if (i < loh_generation)
+        {       
+            if (!reset_only_p)
+            {
+                dprintf (GTC_LOG, ("h%d: SOH: dp:%Id; mo: %Id", heap_number, total_dirtied_pages, total_marked_objects));
+                fire_revisit_event (total_dirtied_pages, total_marked_objects, /* large_objects_p */ FALSE);
+                concurrent_print_time_delta (concurrent_p ? "CR SOH" : "NR SOH");
+                total_dirtied_pages = 0;
+                total_marked_objects = 0;
+            }
+
+            //switch to large segment
+            if (concurrent_p && !reset_only_p)
+            {
+                current_bgc_state = bgc_revisit_uoh;
+            }
+
+            small_object_segments = FALSE;
+            //concurrent_print_time_delta (concurrent_p ? "concurrent marking dirtied pages on SOH" : "nonconcurrent marking dirtied pages on SOH");
+
+            dprintf (3, ("now revisiting large object segments"));
+        }
+        else
+        {
+            if (reset_only_p)
+            {
+                dprintf (GTC_LOG, ("h%d: tdp: %Id", heap_number, total_dirtied_pages));
+            }
+            else
+            {
+                dprintf (GTC_LOG, ("h%d: LOH: dp:%Id; mo: %Id", heap_number, total_dirtied_pages, total_marked_objects));
+                fire_revisit_event (total_dirtied_pages, total_marked_objects, /* large_objects_p */ TRUE);
```

Comment:
there are some comments in this function that talk about "large objects" that will need to be changed.
also another occurence where this ETW event will need to take a generation instead of TRUE/FALSE.

Reply:
the ETW will only support loh in this change. we are trying to not get far from equivalent code, if possible, since the goal is refactoring. 

---
In reply to: [362301895](https://github.com/dotnet/runtime/pull/235#discussion_r362301895) [](ancestors = 362301895)